src/compiler/execution.jl

```diff
@@ -197,15 +197,38 @@ end
 
 ## host-side kernels
 
-struct HostKernel{F,TT} <: AbstractKernel{F,TT}
+mutable struct HostKernel{F,TT} <: AbstractKernel{F,TT}
     ctx::CuContext
     mod::CuModule
     fun::CuFunction
+    uses_random_state::Bool
+    random_state::Union{Nothing,Mem.DeviceBuffer}
+    function HostKernel{F,TT}(ctx::CuContext, mod::CuModule, fun::CuFunction, uses_random_state::Bool) where {F,TT}
+        kernel = new{F,TT}(ctx, mod, fun, uses_random_state, nothing)
+        finalizer(kernel) do k
+            if !isnothing(k.random_state)
+                Mem.free(k.random_state)
+            end
+        end
+        return kernel
+    end
 end
 
 @doc (@doc AbstractKernel) HostKernel
 
-@inline function cudacall(kernel::HostKernel, tt, args...; config=nothing, kwargs...)
+@inline function cudacall(kernel::HostKernel, tt, args...; rng_seed=nothing, config=nothing, kwargs...)
+    if kernel.uses_random_state
+        if !isnothing(kernel.random_state)
+            Mem.free(kernel.random_state)
+        end
+        
+        kws = Dict(kwargs)
+        num_threads = get(kws, :threads, 1) * get(kws, :blocks, 1)
+
+        random_state = init_random_state!(kernel.mod, num_threads, rng_seed)
```

Comment:
The free and re-initialization here should be adjusted to only re-initialize if the state is insufficient, otherwise we can just reuse state from previous launches.